erts/emulator/nifs/common/prim_socket_nif.c

```diff
@@ -16198,8 +16202,83 @@ ERL_NIF_TERM esock_cancel_mode_select(ErlNifEnv*       env,
 }
 #endif // if !defined(__WIN32__)
 
+static
+ERL_NIF_TERM nif_getprotobyname(ErlNifEnv* env,
+        int argc,
+        const ERL_NIF_TERM argv[])
+{
+#if defined(__WIN32__)
+    return enif_raise_exception(env, MKA(env, "notsup"));
+#else
+    ERL_NIF_TERM result;
+
+    char name[1024];
+
+    if (GET_STR(env, argv[0], name, 1023) <= 0) {
+        return enif_make_badarg(env);
+    }
+
+    struct protoent *entry = getprotobyname(name);
+
+    if (entry)
+        return esock_make_protoent(env, entry);
+    else
+        return esock_atom_unknown;
+#endif // defined(__WIN32__)
+}
+
+static
+ERL_NIF_TERM nif_getprotobynumber(ErlNifEnv* env,
+        int argc,
+        const ERL_NIF_TERM argv[])
+{
+#if defined(__WIN32__)
+    return enif_raise_exception(env, MKA(env, "notsup"));
+#else
+    ERL_NIF_TERM result;
+
+    int num;
+
+    if (GET_INT(env, argv[0], &num) <= 0) {
+        return enif_make_badarg(env);
+    }
+
+    struct protoent *entry = getprotobyname(num);
 
+    if (entry)
+        return esock_make_protoent(env, entry);
+    else
+        return esock_atom_unknown;
+#endif // defined(__WIN32__)
+}
 
+#if !defined(__WIN32__)
+static
+ERL_NIF_TERM esock_make_protoent(ErlNifEnv* env,
+        struct protoent *entry)
+{
+    ERL_NIF_TERM name = MKS(env, entry->p_name);
+    ERL_NIF_TERM aliases = MKEL(env);
+
+    for (char *alias = entry->p_aliases; alias; ++alias) {
+        ERL_NIF_TERM item = MKS(env, alias);
+        aliases = enif_make_list_cell(env, item, aliases);
+    }
+
+    ERL_NIF_TERM protocol = MKI(env, entry->p_proto);
```

Comment:
Question is whether it shouldn't be a tuple in form of `{raw, Num}` to make it simpler for use together with `socket:open/3`.